### DISCUSSION TITLE
OPA SDK features

### DISCUSSION THREAD

**Discussion by humbertoc-silva** [open-policy-agent/community#269] [OPA and Rego] [answered]
Hi everyone,

I am testing the OPA SDK, and I would like to know if is it possible to use the following parameters as we use with the OPA server when we run a query?
 
- provenance
- explain
- metrics
- instrument
- strict-builtin-errors

Thank you.

**Comment by srenatus** [comment, marked as answer]
I don't think it's possible to get these as a result of calling `Decision()`. The SDK doesn't expose the same API as the HTTP APIs, which let you set those things; but instead deals with setting up a runtime that is exactly the same as what the HTTP APIs use.

However, it allows setting up plugins for decision logs and status reports -- I suppose some of what you're asking for here could be dealt with in that manner? The decision logs should have information about each and every evaluation, including provenance and metrics.

I don't think anything related to strict-builtin-errors, instrumentation, or explain (traces) is supported, though. Feel free to open an issue if you need a concrete enhancement here.

**Reply by humbertoc-silva** [reply]
Indeed some information I can get from the status and decision log, but the last items you cited are not possible and they are valuable to us, e.g, to permit our teams to turn on the strict-built-in-error is a common use case here.

I will file a new issue about this, thank you!

**Comment by humbertoc-silva** [comment]
@srenatus I filled a new issue related to this discussion and I would like to help implement it if the issue is accepted.

**Comment by charlieegan3** [comment]
This functionality is now available in the SDK following the merging of: https://github.com/open-policy-agent/opa/pull/5438, https://github.com/open-policy-agent/opa/pull/5447, https://github.com/open-policy-agent/opa/pull/5459 and https://github.com/open-policy-agent/opa/pull/5460

The functionality is used by configuring parameters to opa.Decision/opa.Partial calls.

**Reply by humbertoc-silva** [reply]
Thank you @charlieegan3.

**Reply by charlieegan3** [reply]
No probs! Would be interesting to learn more about your use case if you do go this route with the SDK. Keep use posted ✉️

**Reply by humbertoc-silva** [reply]
I am thinking to develop an AWS lambda authorizer using the OPA SDK and these features will be valuable to me. I will give you more details as soon as possible.

**Reply by charlieegan3** [reply]
Great, that sounds interesting! Might be a good project for https://www.openpolicyagent.org/docs/latest/ecosystem/ in time, so do let us know how it goes!

**Reply by anderseknert** [reply]
@humbertoc-silva this might be useful for you https://blog.openpolicyagent.org/serverless-policy-enforcement-connecting-opa-and-aws-lambda-e624f7176a3

**Reply by humbertoc-silva** [reply]
Thank you @anderseknert, it will definitely be helpful.
